File: backend/update_service.py
import json
import os
import re
import shutil
import subprocess
import sys
import threading
import time
import zipfile
from datetime import datetime

import requests

from utils.data_store import DATA_DIR


ROOT_DIR = os.getcwd()
VERSION_FILE = "version.json"
DEFAULT_VERSION = {
    "version": "0.0.0-dev",
    "repo": "Zilvren/Learning-Assitant",
    "asset_name": "Tracker.zip",
    "app_exe": "Tracker.exe",
}
DATA_BACKUP_FILES = {"errors.json", "subjects.json", "config.json", "knowledge.json"}


class UpdateError(Exception):
    pass


def _root_path(*parts):
    return os.path.join(ROOT_DIR, *parts)


def _exe_root():
    if getattr(sys, "frozen", False):
        return os.path.dirname(sys.executable)
    return ROOT_DIR


def _data_path(*parts):
    os.makedirs(DATA_DIR, exist_ok=True)
    return os.path.join(DATA_DIR, *parts)


def _normalize_version(value):
    return str(value or "").strip().lstrip("vV")


def _version_key(value):
    text = _normalize_version(value)
    if re.fullmatch(r"\d{4}\.\d{2}\.\d{2}-\d{4}", text):
        date_part, time_part = text.split("-", 1)
        year, month, day = (int(x) for x in date_part.split("."))
        hour = int(time_part[:2])
        minute = int(time_part[2:])
        return (2, year, month, day, hour, minute)

    if re.fullmatch(r"\d+(\.\d+)*", text):
        parts = [int(x) for x in text.split(".")]
        while len(parts) < 6:
            parts.append(0)
        return (1, *parts[:6])

    return (0, text)


def compare_versions(left, right):
    left_key = _version_key(left)
    right_key = _version_key(right)
    return (left_key > right_key) - (left_key < right_key)


def load_version_info():
    info = DEFAULT_VERSION.copy()
    candidates = []
    if getattr(sys, "frozen", False):
        candidates.append(os.path.join(_exe_root(), VERSION_FILE))
    candidates.append(_root_path(VERSION_FILE))
    candidates.append(os.path.join(DATA_DIR, VERSION_FILE))
    for path in candidates:
        if os.path.isfile(path):
            with open(path, "r", encoding="utf-8-sig") as f:
                loaded = json.load(f)
            if isinstance(loaded, dict):
                info.update({k: v for k, v in loaded.items() if v})
                break
    updater_path = _root_path("Updater.exe")
    info["can_auto_update"] = bool(getattr(sys, "frozen", False) and os.path.isfile(updater_path))
    info["updater_path"] = updater_path if info["can_auto_update"] else ""
    return info


def get_version_response():
    info = load_version_info()
    return {
        "version": info["version"],
        "repo": info["repo"],
        "asset_name": info["asset_name"],
        "app_exe": info["app_exe"],
        "can_auto_update": info["can_auto_update"],
    }


def fetch_latest_release():
    info = load_version_info()
    url = f"https://api.github.com/repos/{info['repo']}/releases/latest"
    try:
        resp = requests.get(url, timeout=12, headers={"Accept": "application/vnd.github+json"})
        resp.raise_for_status()
    except requests.RequestException as exc:
        raise UpdateError(f"检查更新失败：{exc}") from exc

    release = resp.json()
    assets = release.get("assets") or []
    asset = next((item for item in assets if item.get("name") == info["asset_name"]), None)
    latest_version = _normalize_version(release.get("tag_name"))
    current_version = _normalize_version(info["version"])
    result = {
        "current_version": info["version"],
        "latest_version": latest_version,
        "tag_name": release.get("tag_name", ""),
        "has_update": bool(latest_version and compare_versions(latest_version, current_version) > 0),
        "repo": info["repo"],
        "asset_name": info["asset_name"],
        "asset_found": bool(asset),
        "asset_size": asset.get("size", 0) if asset else 0,
        "published_at": release.get("published_at", ""),
        "html_url": release.get("html_url", ""),
        "notes": release.get("body", "") or "",
        "can_auto_update": info["can_auto_update"],
        "download_url": asset.get("browser_download_url", "") if asset else "",
    }
    return result


def check_update(force=False):
    try:
        release = fetch_latest_release()
        release["ok"] = True
        release["checked_at"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        release.pop("download_url", None)
        return release
    except UpdateError as exc:
        return {
            "ok": False,
            "message": str(exc),
            "checked_at": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            **get_version_response(),
        }


def save_pre_update_snapshot():
    backup_dir = _data_path("backups")
    os.makedirs(backup_dir, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    snapshot = os.path.join(backup_dir, f"pre-update-{stamp}.zip")
    with zipfile.ZipFile(snapshot, "w", zipfile.ZIP_DEFLATED) as zf:
        for filename in sorted(DATA_BACKUP_FILES):
            path = _data_path(filename)
            if os.path.isfile(path):
                zf.write(path, arcname=filename)
    return snapshot


def download_update_package(release):
    if not release.get("download_url"):
        raise UpdateError("最新 Release 中没有找到 Tracker.zip")
    updates_dir = _data_path("updates")
    os.makedirs(updates_dir, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    target = os.path.join(updates_dir, f"Tracker-{release['latest_version']}-{stamp}.zip")
    tmp = target + ".download"

    try:
        with requests.get(release["download_url"], stream=True, timeout=30) as resp:
            resp.raise_for_status()
            with open(tmp, "wb") as f:
                for chunk in resp.iter_content(chunk_size=1024 * 256):
                    if chunk:
                        f.write(chunk)
        os.replace(tmp, target)
    except requests.RequestException as exc:
        if os.path.exists(tmp):
            os.remove(tmp)
        raise UpdateError(f"下载更新失败：{exc}") from exc

    if not zipfile.is_zipfile(target):
        os.remove(target)
        raise UpdateError("下载的更新包不是有效 zip 文件")
    return target


def _exit_later(delay=1.0):
    def _stop():
        time.sleep(delay)
        os._exit(0)

    threading.Thread(target=_stop, daemon=True).start()


def apply_update():
    info = load_version_info()
    if not info["can_auto_update"]:
        raise UpdateError("当前运行环境不支持自动替换，请使用打包后的 Tracker.exe")

    release = fetch_latest_release()
    if compare_versions(release["latest_version"], info["version"]) <= 0:
        return {"message": "当前已是最新版本", **check_update(force=True)}
    if not release["asset_found"]:
        raise UpdateError(f"最新 Release 中没有找到 {info['asset_name']}")

    snapshot = save_pre_update_snapshot()
    package = download_update_package(release)
    updater = info["updater_path"]
    app_dir = ROOT_DIR

    cmd = [
        updater,
        "--package",
        package,
        "--app-dir",
        app_dir,
        "--app-exe",
        info["app_exe"],
        "--pid",
        str(os.getpid()),
    ]
    creationflags = 0
    if os.name == "nt":
        creationflags = subprocess.CREATE_NEW_PROCESS_GROUP | subprocess.DETACHED_PROCESS
    subprocess.Popen(cmd, cwd=app_dir, close_fds=True, creationflags=creationflags)
    _exit_later()

    return {
        "message": "更新包已下载，程序即将重启并安装更新",
        "latest_version": release["latest_version"],
        "package": package,
        "snapshot": os.path.basename(snapshot),
    }

File: service/update_service.go
package service

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"study-tracker-go/store"
)

const versionFile = "version.json"

var dataBackupFiles = []string{"config.json", "errors.json", "knowledge.json", "subjects.json"}

type UpdateApplyError struct {
	Message string
}

func (e UpdateApplyError) Error() string {
	return e.Message
}

type VersionInfo struct {
	Version   string `json:"version"`
	Repo      string `json:"repo"`
	AssetName string `json:"asset_name"`
	AppExe    string `json:"app_exe"`

	AppDir        string `json:"-"`
	CanAutoUpdate bool   `json:"-"`
	UpdaterPath   string `json:"-"`
}

type releaseInfo struct {
	CurrentVersion string
	LatestVersion  string
	TagName        string
	HasUpdate      bool
	Repo           string
	AssetName      string
	AssetFound     bool
	AssetSize      int64
	PublishedAt    string
	HTMLURL        string
	Notes          string
	CanAutoUpdate  bool
	DownloadURL    string
}

type githubReleaseResponse struct {
	TagName     string `json:"tag_name"`
	PublishedAt string `json:"published_at"`
	HTMLURL     string `json:"html_url"`
	Body        string `json:"body"`
	Assets      []struct {
		Name               string `json:"name"`
		Size               int64  `json:"size"`
		BrowserDownloadURL string `json:"browser_download_url"`
	} `json:"assets"`
}

func GetVersionResponse() map[string]interface{} {
	info := loadVersionInfo()
	return map[string]interface{}{
		"version":         info.Version,
		"repo":            info.Repo,
		"asset_name":      info.AssetName,
		"app_exe":         info.AppExe,
		"can_auto_update": info.CanAutoUpdate,
	}
}

func CheckUpdate(force bool) map[string]interface{} {
	_ = force
	checkedAt := time.Now().Format("2006-01-02 15:04:05")
	release, err := fetchLatestRelease()
	if err != nil {
		response := GetVersionResponse()
		response["ok"] = false
		response["message"] = "检查更新失败：" + err.Error()
		response["checked_at"] = checkedAt
		return response
	}

	result := releaseToMap(release)
	result["ok"] = true
	result["checked_at"] = checkedAt
	delete(result, "download_url")
	return result
}

func ApplyUpdate() (map[string]interface{}, error) {
	info := loadVersionInfo()
	if !info.CanAutoUpdate {
		return nil, UpdateApplyError{Message: "当前运行环境不支持自动替换，请使用打包后的 Tracker.exe"}
	}

	release, err := fetchLatestRelease()
	if err != nil {
		return nil, UpdateApplyError{Message: "检查更新失败：" + err.Error()}
	}
	if compareVersions(release.LatestVersion, info.Version) <= 0 {
		result := releaseToMap(release)
		result["message"] = "当前已是最新版本"
		result["ok"] = true
		delete(result, "download_url")
		return result, nil
	}
	if !release.AssetFound || strings.TrimSpace(release.DownloadURL) == "" {
		return nil, UpdateApplyError{Message: fmt.Sprintf("最新 Release 中没有找到 %s", info.AssetName)}
	}

	snapshot, err := savePreUpdateSnapshot()
	if err != nil {
		return nil, err
	}
	packagePath, err := downloadUpdatePackage(release)
	if err != nil {
		return nil, UpdateApplyError{Message: "下载更新失败：" + err.Error()}
	}
	updaterRunPath, err := copyUpdaterForRun(info.UpdaterPath)
	if err != nil {
		return nil, err
	}
	if err := startUpdater(updaterRunPath, packagePath, info.AppDir, info.AppExe); err != nil {
		return nil, err
	}

	go func() {
		time.Sleep(time.Second)
		os.Exit(0)
	}()

	return map[string]interface{}{
		"message":        "更新包已下载，程序即将重启并安装更新",
		"latest_version": release.LatestVersion,
		"package":        packagePath,
		"snapshot":       filepath.Base(snapshot),
	}, nil
}

func loadVersionInfo() VersionInfo {
	appDir := appDirectory()
	info := VersionInfo{
		Version:   "0.0.0-dev",
		Repo:      "Zilvren/Learning-Assitant",
		AssetName: "Tracker.zip",
		AppExe:    "Tracker.exe",
		AppDir:    appDir,
	}

	if data, err := os.ReadFile(filepath.Join(appDir, versionFile)); err == nil {
		_ = json.Unmarshal(data, &info)
	}
	normalizeVersionInfo(&info)
	info.AppDir = appDir
	info.UpdaterPath = filepath.Join(appDir, "Updater.exe")
	info.CanAutoUpdate = fileExists(info.UpdaterPath)
	return info
}

func normalizeVersionInfo(info *VersionInfo) {
	if strings.TrimSpace(info.Version) == "" {
		info.Version = "0.0.0-dev"
	}
	if strings.TrimSpace(info.Repo) == "" {
		info.Repo = "Zilvren/Learning-Assitant"
	}
	if strings.TrimSpace(info.AssetName) == "" {
		info.AssetName = "Tracker.zip"
	}
	if strings.TrimSpace(info.AppExe) == "" {
		info.AppExe = "Tracker.exe"
	}
}

func appDirectory() string {
	if cwd, err := os.Getwd(); err == nil {
		return cwd
	}
	return "."
}

func fetchLatestRelease() (releaseInfo, error) {
	info := loadVersionInfo()
	url := "https://api.github.com/repos/" + info.Repo + "/releases/latest"
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return releaseInfo{}, err
	}
	req.Header.Set("Accept", "application/vnd.github+json")

	client := &http.Client{Timeout: 12 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return releaseInfo{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return releaseInfo{}, fmt.Errorf("GitHub 返回 HTTP %d", resp.StatusCode)
	}

	var payload githubReleaseResponse
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return releaseInfo{}, err
	}
	return parseReleasePayload(info, payload), nil
}

func parseReleasePayload(info VersionInfo, payload githubReleaseResponse) releaseInfo {
	latest := normalizeVersion(payload.TagName)
	result := releaseInfo{
		CurrentVersion: info.Version,
		LatestVersion:  latest,
		TagName:        payload.TagName,
		HasUpdate:      latest != "" && compareVersions(latest, info.Version) > 0,
		Repo:           info.Repo,
		AssetName:      info.AssetName,
		PublishedAt:    payload.PublishedAt,
		HTMLURL:        payload.HTMLURL,
		Notes:          payload.Body,
		CanAutoUpdate:  info.CanAutoUpdate,
	}
	for _, asset := range payload.Assets {
		if asset.Name == info.AssetName {
			result.AssetFound = true
			result.AssetSize = asset.Size
			result.DownloadURL = asset.BrowserDownloadURL
			break
		}
	}
	return result
}

func releaseToMap(release releaseInfo) map[string]interface{} {
	return map[string]interface{}{
		"current_version": release.CurrentVersion,
		"latest_version":  release.LatestVersion,
		"tag_name":        release.TagName,
		"has_update":      release.HasUpdate,
		"repo":            release.Repo,
		"asset_name":      release.AssetName,
		"asset_found":     release.AssetFound,
		"asset_size":      release.AssetSize,
		"published_at":    release.PublishedAt,
		"html_url":        release.HTMLURL,
		"notes":           release.Notes,
		"can_auto_update": release.CanAutoUpdate,
		"download_url":    release.DownloadURL,
	}
}

func savePreUpdateSnapshot() (string, error) {
	backupDir := filepath.Join(store.DataDir(), "backups")
	if err := os.MkdirAll(backupDir, 0755); err != nil {
		return "", err
	}
	snapshot := filepath.Join(backupDir, "pre-update-"+time.Now().Format("20060102-150405")+".zip")
	out, err := os.Create(snapshot)
	if err != nil {
		return "", err
	}
	defer out.Close()

	zw := zip.NewWriter(out)
	for _, name := range dataBackupFiles {
		path := store.Path(name)
		if !fileExists(path) {
			continue
		}
		if err := addFileToZip(zw, path, name); err != nil {
			_ = zw.Close()
			return "", err
		}
	}
	return snapshot, zw.Close()
}

func addFileToZip(zw *zip.Writer, source string, name string) error {
	src, err := os.Open(source)
	if err != nil {
		return err
	}
	defer src.Close()

	dst, err := zw.Create(name)
	if err != nil {
		return err
	}
	_, err = io.Copy(dst, src)
	return err
}

func downloadUpdatePackage(release releaseInfo) (string, error) {
	updatesDir := filepath.Join(store.DataDir(), "updates")
	if err := os.MkdirAll(updatesDir, 0755); err != nil {
		return "", err
	}
	stamp := time.Now().Format("20060102-150405")
	name := fmt.Sprintf("Tracker-%s-%s.zip", safeFilenamePart(release.LatestVersion), stamp)
	target := filepath.Join(updatesDir, name)
	tmp := target + ".download"

	req, err := http.NewRequest(http.MethodGet, release.DownloadURL, nil)
	if err != nil {
		return "", err
	}
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		_ = os.Remove(tmp)
		return "", fmt.Errorf("下载地址返回 HTTP %d", resp.StatusCode)
	}

	out, err := os.Create(tmp)
	if err != nil {
		return "", err
	}
	if _, err := io.Copy(out, resp.Body); err != nil {
		_ = out.Close()
		_ = os.Remove(tmp)
		return "", err
	}
	if err := out.Close(); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	if err := os.Rename(tmp, target); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	if err := validateZip(target); err != nil {
		_ = os.Remove(target)
		return "", err
	}
	return target, nil
}

func copyUpdaterForRun(updaterPath string) (string, error) {
	updatesDir := filepath.Join(store.DataDir(), "updates")
	if err := os.MkdirAll(updatesDir, 0755); err != nil {
		return "", err
	}
	target := filepath.Join(updatesDir, "updater-run-"+time.Now().Format("20060102-150405")+".exe")
	src, err := os.Open(updaterPath)
	if err != nil {
		return "", err
	}
	defer src.Close()
	dst, err := os.Create(target)
	if err != nil {
		return "", err
	}
	if _, err := io.Copy(dst, src); err != nil {
		_ = dst.Close()
		return "", err
	}
	if err := dst.Close(); err != nil {
		return "", err
	}
	return target, nil
}

func startUpdater(updaterPath, packagePath, appDir, appExe string) error {
	cmd := exec.Command(
		updaterPath,
		"--package", packagePath,
		"--app-dir", appDir,
		"--app-exe", appExe,
		"--pid", strconv.Itoa(os.Getpid()),
	)
	cmd.Dir = appDir
	if err := cmd.Start(); err != nil {
		return err
	}
	return cmd.Process.Release()
}

func validateZip(path string) error {
	reader, err := zip.OpenReader(path)
	if err != nil {
		return fmt.Errorf("下载的更新包不是有效 zip 文件")
	}
	return reader.Close()
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func safeFilenamePart(text string) string {
	text = strings.TrimSpace(text)
	if text == "" {
		return "unknown"
	}
	re := regexp.MustCompile(`[^A-Za-z0-9._-]+`)
	return re.ReplaceAllString(text, "_")
}

func normalizeVersion(value string) string {
	return strings.TrimLeft(strings.TrimSpace(value), "vV")
}

func compareVersions(left, right string) int {
	left = normalizeVersion(left)
	right = normalizeVersion(right)
	leftKind, leftInts, leftText := versionKey(left)
	rightKind, rightInts, rightText := versionKey(right)
	if leftKind != rightKind {
		return compareInt(leftKind, rightKind)
	}
	if leftKind == 0 {
		return strings.Compare(leftText, rightText)
	}
	for i := 0; i < len(leftInts) && i < len(rightInts); i++ {
		if c := compareInt(leftInts[i], rightInts[i]); c != 0 {
			return c
		}
	}
	return compareInt(len(leftInts), len(rightInts))
}

func versionKey(value string) (int, []int, string) {
	dateRe := regexp.MustCompile(`^\d{4}\.\d{2}\.\d{2}-\d{4}$`)
	if dateRe.MatchString(value) {
		parts := []int{}
		for _, piece := range regexp.MustCompile(`[.-]`).Split(value, -1) {
			number, _ := strconv.Atoi(piece)
			if len(piece) == 4 && len(parts) == 3 {
				hour, _ := strconv.Atoi(piece[:2])
				minute, _ := strconv.Atoi(piece[2:])
				parts = append(parts, hour, minute)
				continue
			}
			parts = append(parts, number)
		}
		return 2, parts, ""
	}

	numericRe := regexp.MustCompile(`^\d+(\.\d+)*$`)
	if numericRe.MatchString(value) {
		pieces := strings.Split(value, ".")
		parts := make([]int, 0, 6)
		for _, piece := range pieces {
			number, _ := strconv.Atoi(piece)
			parts = append(parts, number)
		}
		for len(parts) < 6 {
			parts = append(parts, 0)
		}
		return 1, parts[:6], ""
	}

	return 0, nil, value
}

func compareInt(left, right int) int {
	switch {
	case left > right:
		return 1
	case left < right:
		return -1
	default:
		return 0
	}
}
